test/test_common/network_utility.h

```diff
@@ -1,13 +1,34 @@
 #pragma once
 
+#include "envoy/network/address.h"
+
 namespace Network {
 namespace Test {
 
 /**
- * @returns an unused port, either be getting one from a port server, or by searching
- * through the ports for an unused port.
+ * Determines if the passed in address and port is available for binding. If the port is zero,
+ * the OS should pick an unused port for the supplied address (e.g. for the loopback address).
+ * NOTE: this is racy, as it does not provide a means to keep the port reserved for the
+ * caller's use.
+ * @param addrPort a valid host address (e.g. an address of one of the network interfaces
+ *        of this host, or the any address or the loopback address) and port (zero to indicate
+ *        that the OS should pick an unused address.
+ * @param type the type of socket to be tested.
+ * @returns the address and port (selected if zero was the passed in port) that can be used for
+ *          listening, else nullptr if the address and port are not free.
+ */
+Address::InstanceConstSharedPtr checkPortAvailability(Address::InstanceConstSharedPtr addrPort,
```

Comment:
I think this was already commented on, but FWIW checkPortAvailability() seems like not a great name since in the zero port case we actually assign a port. Not sure what would be better though. checkAndMaybeAssignPort() ?